### ISSUE TITLE
Feature Request: Workflow to create branch from issue

### ISSUE THREAD

**Issue by krnlde** [isaacs/github#1125]
Coming from Jira I really miss the possibility to create a [feature|hotfix] branch that matches the current issue with its name and ID. I suggest to put this workflow in the sidebar of an issue where the Assignees, Labels and Milestones reside.

The following workflow would be nice:
1. Create an issue "My Example Issue" with id 42
2. See a "Create Branch for Issue"-button in the sidebar
3. Click on it and choose where to branch from (e.g. `develop`)
4. define the branch-type according to gitflow (feature/hotfix)
5. Create the branch whose name will be inferred from the issue's information `feature/my-example-issue-42` or choose your own name
6. `git fetch` on your local machine and see the new branch
7. switch to it, make the necessary changes described in the issue
8. See the number of commits to this issue in the sidebar
9. Create a pull-request for the branch directly from the sidebar

**Comment by rorybot** [comment]
I'm astounded this doesn't exist already.

**Comment by rstockm** [comment]
And tie it to the "Projects" board, as well. There are so many loose ends on GitHub in terms of workflow.

**Comment by galvani** [comment]
I have just spend 30 minutes searching how to do it :-) I thought I am dumb.

**Comment by andrewslavin** [comment]
@galvani +1.

I raised an issue on an open-source project, the author accepted it as a bug and asked if I could contribute by raising a PR. Coming from Jira/Bitbucket experience like @krnlde, I spend 30 min looking for a way to start a branch from my issue, and eventually ended up here...

**Comment by shubhamkulkarni** [comment]
I come from JIRA as well and use the same workflow as mentioned by  @krnlde, which is much efficient. 
How can we take this issue forward and request for the same?

**Comment by paesku** [comment]
used to gitlab every day, it surprised me to end up in this thread 👀 .

**Comment by kellymaker** [comment]
+1 on this.  May be tricky as a Project is only loosely coupled....

**Comment by jkrenge** [comment]
+1

**Comment by gs377209** [comment]
+1

**Comment by krnlde** [comment]
Please stop spamming this issue with +1s. There is a thumbsup/-down feature for this if you really like to share your interest.

**Comment by Jambazishvili** [comment]
I needed kinda same behavior described, and recently came up with solution which may partially solve the problem, In addition I want to note that this no totally/exactly solves the problem originally described, but it may **partially** do it:

1. Open issue ("#23" for example)
2. Locally create branch from desired branch, as usual I name branch i{issue_number} (i23)
3. Once, i've done fixes for issue I create pull request and in the pull request body write "fixes #23"
4. This automatically closes "issue #23"

this process is described in this [post](https://blog.github.com/2013-05-14-closing-issues-via-pull-requests/) and really helped me to partially automate such workflow. hope it helps you too

**Comment by maxdmayhew** [comment]
Please!

**Comment by maxdmayhew** [comment]
Please

and include it in enterprise!

**Comment by maxdmayhew** [comment]
Please!

and include it in enterprise!

**Comment by MichaelMausbach** [comment]
Coming from gitlab, I couldn't believe that this function is not included in github. I would be great to have it.

**Comment by KevinMind** [comment]
Please please please.

**Comment by KevinMind** [comment]
Though. Someone could use the api to build this feature..... ;) See netlify cms.  Any takers?

**Comment by Jambazishvili** [comment]
@KevinMind can you suggest some starting point?

**Comment by krnlde** [comment]
More specific: How do you integrate UI elements in the sidepanel to have a button that triggers the API?

**Comment by othreecodes** [comment]
This doesn't exist yet?
:disappointed:

**Comment by galvani** [comment]
it does not

**Comment by stritti** [comment]
every time I search this feature again ...

**Comment by klaussemarsse** [comment]
tried the same and ended up here.  It would be really a good one feature

**Comment by magick93** [comment]
Gitlab FTW.

Github - time to up your game!

**Comment by zteater** [comment]
This is OOTB functionality with your competitors (GitLab, BitBucket + Jira, VersionOne).

**Comment by sinelga** [comment]
It must be done! :)

**Comment by mtumilowicz** [comment]
Up, up, up, up!

**Comment by btbam91** [comment]
Please do this!

**Comment by siulcode** [comment]
I am shocked by knowing that this feature does not exist.

**Comment by mpittkin** [comment]
+1
I'm surprised by how much I really, REALLY miss this feature after Jira.

**Comment by pwujek** [comment]
+1

**Comment by BenjAzko** [comment]
+1

**Comment by siulcode** [comment]
Do we need 10,000 signatures for this one to get attention? 😅

**Comment by junwang-stratus** [comment]
Please make this Github great again!

**Comment by alx-andru** [comment]
+1

**Comment by joelchretien** [comment]
+1

**Comment by martinvelden** [comment]
Would perfectly fit into the automation of the Project Boards as well.

New issue -> To do
Create Branch from issue -> In progress
Close issue -> Done

This and letting labels interact with the boards (set column, how high up in priority) are the only things I'm missing.

**Comment by mpittkin** [comment]
That’s a really great idea. That would make it possible to do some really neat things like story mapping. The top level stories with one label, then the next level clearly separated vertically, then so on. If you can then select issues by label and create a release out of those issues there’s your whole story mapping workflow.

That might be a nice feature too, creating a release based on a label, I know that’s kind of what a release is already, but that would fit in with the user story mapping approach nicely.

**Comment by andrewmcallister** [comment]
+1

**Comment by raditotev** [comment]
+1

**Comment by AkshayBudhiraja** [comment]
We could really use this functionality in our team's workflow as well, as I imagine other teams can as well. Coming from JIRA, this is the one thing that we are missing.

**Comment by once-returner** [comment]
Being able to create a branch from an issue (and a branch from an issue on the project board) is behavior that I expected would be there by default, this should definitely be prioritized.

**Comment by adumont** [comment]
Over 1 yr after this feature is still not implemented?? weird, it sounds like a basic feature.

**Comment by galvani** [comment]
it's been labeled a month ago, it seems there is coming some progress now.

**Comment by zpalazov** [comment]
+1

**Comment by johnfoland** [comment]
+1

**Comment by CraigLager** [comment]
+1 - we've ended up implementing this through a webhook when a label gets applied.

**Comment by garettridge** [comment]
+1

**Comment by VenkateshKoka** [comment]
@CraigLager Can you explain how you didi it.

**Comment by CraigLager** [comment]
We have a webhook which basically calls the `IssueLabelled()` function in this:

https://gist.github.com/CraigLager/f8882c9dcc71d4b85aeb074f36733b08

It relies on Octokit.net for the GitClient object

**Comment by VenkateshKoka** [comment]
thanks Craig for the reply.

**Comment by AtifChatdroid** [comment]
+1

**Comment by TPS** [comment]
@clarkbw Some implemented improvements dance right around this issue, but none actually does _this_.…  Any news?  E.g.:
- https://github.blog/2013-05-14-closing-issues-via-pull-requests/
- https://github.blog/changelog/2018-10-10-open-issue-from-review-comment/

**Comment by clarkbw** [comment]
Nothing yet but I would like to make this happen. 

You could use an Action to listen for an issue label / assign event and then crest the PR.

**Comment by obar1** [comment]
too similar to gitlab that s why they dont want to do it ... [my opinion]

**Comment by Antonhansel** [comment]
![image](https://user-images.githubusercontent.com/4123811/54759148-ccfd5e80-4bed-11e9-80c3-b5089d7a05d5.png)

**Comment by avatsaev** [comment]
So? what's up?

**Comment by pi-mont** [comment]
Hey Microsoft, throw some cash at this issue!

**Comment by jakubsuchybio** [comment]
I think they are already working on that with high priority.
In Visual Studio Release Event I asked via twitter if it is already possible to create new branch from Work Item in Azure Devops when the repo is hosted on GitHub, they answered that for now it is only posible for repos hosted in Azure Devops, but they are working on it for GitHub as well with high priority, because a lot of ppl is requesting it. ( Maybe this issue :D ) By that I suspect that they need to add that to the GitHub API and when they do implement it for Azure Devops I think they will co-release it in Github too.

**Comment by robertofrega** [comment]
Hi guys.

Is this any close to what the feature could be as an MVF?

**Comment by Ruegen** [comment]
A
Another request here, this would be a great feature to have

**Comment by zsjenei** [comment]
May 2019 and still wishing this feature :)

**Comment by Tillman32** [comment]
+1 this is a MUST.

**Comment by DaBozUK** [comment]
Gitlab.com provides this functionality and more with your github user in their free tier.
https://about.gitlab.com/pricing/gitlab-com/feature-comparison/

**Comment by NikoMontana** [comment]
common Microsoft..

**Comment by robvanderleek** [comment]
I've written a small app that creates a branch (e.g. `issue-15-Fix_nasty_bug`) after an issue is assigned. Not in the app listing yet but you can already install it: https://github.com/apps/create-issue-branch (feedback is appreciated, please create an [issue](https://github.com/robvanderleek/create-issue-branch/issues))

**Comment by suanmeiguo** [comment]
Saw this issue today so just wrote the app. Finally when I come here to comment found someone already worked on this 😂 . Anyway since I already built it out I'll just post it here. Just more options. Feel free to try: https://github.com/apps/auto-branch.

**Comment by suanmeiguo** [comment]
Hello, just wanna give an update that this is available in [marketplace](https://github.com/marketplace/auto-branch) now.

**Comment by robvanderleek** [comment]
The Create Issue Branch app is also available in the [marketplace](https://github.com/marketplace/create-issue-branch) and in the [Probot directory](https://probot.github.io/apps/create-issue-branch/).

**Comment by Syleron** [comment]
This issue really needs some love..

**Comment by HarisSpahija** [comment]
Can we as a community do something with this issue? Hacktoberfest is uppon is and I think everyone would be interested in chipping in to this!

**Comment by blairg23** [comment]
How is this issues almost 2 years old with no progress?

**Comment by krnlde** [comment]
First of all this is a private issue tracker from Isaac. Second there is not "no progress". Third, check yourself: how can you contribute other than just complaining? Is complaining for free software really the best way to respond?

**Comment by serverhorror** [comment]
Be nice to each other :)
Another way to read this is: “I think it’s kind of sad that this repository exists and GitHub doesn’t seem to take advantage of the data hidden here. How can GitHub not pick up on long-standing issues like this one?” (Insert expression of disappointment towards GitHub and gratitude towards this repo for being able to let some steam off here)

**Comment by farmisen** [comment]
@blairg23, @HarisSpahija: @suanmeiguo has a solution and a working one, I have installed it today and it pretty much address the feature requested here. Try it for yourself: https://github.com/isaacs/github/issues/1125#issuecomment-522397132

**Comment by HarisSpahija** [comment]
@farmisen even though it is a solution, it is not the requested feature. The whole point of branch creating from an issue is to do it trough the UI like Jira does.

**Comment by farmisen** [comment]
@HarisSpahija technically it does it from the ui since it happens only when you assign someone to the issue what happens through the ui but I get your point. I feel @suanmeiguo has done most of the hard work to get there and it should not be difficult at that point to either fork his project to add a create branch button or get inspired by his code to create it from scratch.

**Comment by dalezak** [comment]
In case it helps someone, here are the commands I use on Mac to create a pull request for an issue:
 
```
ISSUE_NAME="123 Name Of The Issue"
BRANCH_NAME=$(echo $ISSUE_NAME | tr '[:upper:]' '[:lower:]' | tr '/' '-' | tr ' ' '-')
git checkout -b $BRANCH_NAME
git add .
git add -u
git commit -m "Description of my changes for issue #123"
git push -u origin $BRANCH_NAME
```

**Comment by akosyakov** [comment]
I wonder whether everybody likes so long branch names?

[Gitpod](https://www.gitpod.io) creates branches automatically if it is started from the issue context, for instance for [this issue](https://gitpod.io#https://github.com/isaacs/github/issues/1125). But currently we use only GH issue number, i.e. it will be `GH-1125`. Looking for the way to improve it. The current idea is something like:
`{username}/{dashedTitle.substring(0,10).lowercase}#{issuenr}`.

What do you think?

**Comment by serverhorror** [comment]
> I wonder whether everybody likes so long branch names?

I like it because it will tell me in the branch name what it is about. I'd prefer something like:
`{labelName}/{issuenr}-{issueTitleSlug.lowercase}`

Where `{labelName}` would be a list of labels that are applied. As in _feature_, _bug_, … and as a default -- without any labels present -- it would be just one of those labels to the issue and then use that as the prefix.

**Comment by landsman** [comment]
Hah, right now my coleague told me that he very missing this feature. I agree!

**Comment by dojo90** [comment]
Based on **dalezak** answer, here is a powershell script

```powershell
param([string]$IssueName)

$branchName = $IssueName.ToLower().Replace('/', '-').Replace(' ', '-')
$branchName = $branchName.Substring('0', '40')

echo "New branch '$branchName' created"

git checkout -b $branchName
git add .
git add -u
#git commit -m "Description of my changes for issue $IssueName"
#git push -u origin $branchName
```

usage:

`./createBranch "1 hello world"`

**Comment by 23tux** [comment]
Could this feature be your New Year's Resolution @github and @microsoft?

**Comment by RobinTHOREL** [comment]
:pray: @microsoft

**Comment by krisjans** [comment]
Pretty please :)

**Comment by noebrito** [comment]
How is this still not a thing?

**Comment by robvanderleek** [comment]
There's a GitHub App/Action that automatically creates branches for issues (for example: it creates a branch called `issue-15-Fix_nasty_bug` for Issue 15: "Fix nasty bug!")
You can install this app for an organization/account/repo [here](https://github.com/apps/create-issue-branch). 
By default the app creates a branch after the issue is assigned but there's also a ChatOps mode where it creats a branch only after commenting `/create-issue-branch` (or `/cib`) on the issue.
There are more configuration options, see the documentation in the [repository](https://github.com/robvanderleek/create-issue-branch).

FD: I'm the author of the app and the functionality it offers is not completely the same as requested by the creator of this issue. However, maybe it's useful for your development workflow for the time being. [Feedback](https://github.com/robvanderleek/create-issue-branch/issues) on the app is very welcome!

**Comment by dmstern** [comment]
It would be also very useful if this feature (connect issue+branch+pr) would be included in the [`hub` CLI](https://hub.github.com/).

**Comment by onacit** [comment]
I'm sorry. I failed to tracking the whole comments.

There's still no way, is it?

**Comment by galangel** [comment]
I also want it,
[I can relate to this](https://github.community/t5/How-to-use-Git-and-GitHub/Github-projects-issues-and-pull-requests/td-p/12367)

**Comment by advra** [comment]
Its been years. Are we ever going to see this feature implemented?

**Comment by blairg23** [comment]
It's built into BitBucket, easy to switch and it's free :)

**Comment by noebrito** [comment]
> It's built into BitBucket, easy to switch and it's free :)

That's a funny joke.

**Comment by MarcvdSluys** [comment]
As a workaround, I got pretty far meeting my needs using two tools:

1. [create-issue-branch](https://github.com/marketplace/actions/create-issue-branch) by @robvanderleek to automatically create a branch with the same name as an issue when assigning the issue to someone.  However, the issue and branch are not truely connected, and when creating a Pull request for the branch, I still need to manually link the issue to it to ensure that the issue is closed with the PR.
2. [Project bot](https://github.com/philschatz/project-bot) to automatically move a Projects Card for an Issue to the **In progress** column when the issue is assigned (and move it back to **To do** when unassigned).  However, this is done with Notes cards that are supposed to stay at the top or bottom of the column, and sometimes seem to slip to the middle somehow.  Also, I have to add these two cards to all projects for them to work.  In addition, they clutter up the columns a bit.

**Comment by robvanderleek** [comment]
Hi @MarcvdSluys 

I've just released a new feature for [create-issue-branch](https://github.com/apps/create-issue-branch): Automatically close issues after a pull request merge

You need to activate this feature in the configuration YAML (and also grant the app additional permissions if you've already installed it), please see the [documentation](https://github.com/robvanderleek/create-issue-branch#automatically-close-issues-after-a-pull-request-merge)

I hope this better supports your development workflow, let me [know](https://github.com/robvanderleek/create-issue-branch/issues) otherwise

**Comment by matheusslg** [comment]
Nothing?

**Comment by sblantipodi** [comment]
ok now it's time to implement it.

**Comment by Vibpositive** [comment]
That is long needed

**Comment by frankbekema** [comment]
+1 Would be awesome

**Comment by MarcvdSluys** [comment]
Note that "my" workaround has been improved by @robvanderleek since I posted it on April 11, and works pretty well.  I use his [create-issue-branch](https://github.com/marketplace/actions/create-issue-branch) to automatically create a branch with the same name as an issue when assigning the issue to someone.  The issue is now also automatically closed when the PR is merged.

This doesn't mean though that a native solution wouldn't be preferable...

**Comment by Choqs** [comment]
What is Github waiting for ?

**Comment by LifeIsStrange** [comment]
It's not even on their roadmap https://github.com/github/roadmap
Are they even aware of this issue?!

**Comment by krnlde** [comment]
You guys are even aware that this is in isaacs account, right?

**Comment by Falnesio** [comment]
People being left behind...

https://blog.axosoft.com/gitkraken-git-gui-v7-2-gitlab-issues/

**Comment by Falnesio** [comment]
https://blog.axosoft.com/gitkraken-git-gui-v7-3-github-issues

If you use GitHub or GitHub Enterprise issue trackers, you can now accomplish the following actions directly from the GitKraken Git GUI:

    View GitHub issues 
    Create filters for GitHub issues 
    View GitHub issue details
    Create branches tied to GitHub issues
    Edit the assignee(s) of a GitHub issue*
    Edit the state of a GitHub issue*
    Add/remove labels on a GitHub issue*
    Comment on a GitHub issue*
    Create new GitHub issues* 

*Paid features.

**Comment by endiesworld** [comment]
# Create github branch from issue.

this is still not a thing as at 21st of Sept 2020.

**Comment by jacobobq** [comment]
# Create github branch from issue
Still not a thing as of 22st of Sept 2020.

**Comment by krnlde** [comment]
Still an issue in the private fork of issac and not github itself... 🧠

**Comment by pilkch** [comment]
Do the github developers use github?

**Comment by TomKJ** [comment]
Dear Github Community the **Create github branch from issue** would make things much easier.
I used also Bitbucket in my last company and was surprised that this feature doesn´t exist here in Github.

**Comment by alexmacniven** [comment]
I don't think this is a feature we're going to see on the web interface.

It has been implemented in the [github extension](https://code.visualstudio.com/docs/editor/github) for VSCode. I think it's far more likely to be implemented in [cli/cli](https://github.com/cli/cli) which is in (very) active development right now.

**Comment by miniscruff** [comment]
For those who have the new [github cli](https://github.com/cli/cli) I made an alias to create a branch locally.
`gh alias set start 'gh issue view $1 | head -n 1 | cut -c8- | tr "[:upper:]" "[:lower:]" | sed "s/ /-/g" | (echo -n $1- && cat) | xargs git checkout -b' --shell` ( note that I am sure this could be simpler )
And then you can just do `gh start <issue number>`.

**Comment by SmaugOfErebor** [comment]
If I wasn't a freeloading user, I would totally complain that this feature doesn't exist.

**Comment by MagicUnderHood** [comment]
I'm still waiting for this feature

**Comment by zteater** [comment]
Now that the CLI is out, it seems like a logical place for this feature
https://cli.github.com/manual/gh_issue

**Comment by n-humphries-gen** [comment]
Company is making the switch from GitLab to GitHub and can't believe this isn't a feature ..

**Comment by cosmikwolf** [comment]
This is my favorite feature of GitLab

**Comment by dioncodes** [comment]
Switched from GitHub to a self-hosted GitLab instance due to this missing feature... I don't understand why they still haven't implemented this feature yet.

**Comment by yuri-norwood** [comment]
Coming from Gitlab and Jira, this is a basic part of my daily workflow, please consider adding it

**Comment by bumbeishvili** [comment]
I just created Tampermonkey script for this feature

Here are the source and installation instructions - https://github.com/bumbeishvili/create-branch-from-issue

![analytics](https://user-images.githubusercontent.com/6873202/103043775-86dde780-4597-11eb-8f8a-08968f3b092c.gif)

**Comment by tjwrona1992** [comment]
I just want to point out that this issue has the most likes out of every other issue on here and is over 3 years old... It should really be getting more attention. This is the most wanted issue by the entire community.

**Comment by mcsantiago** [comment]
Wait, this is not a thing yet? I got so used to linking my branches in Jira that it did not occur to me that Github hasn't caught up.

**Comment by bumbeishvili** [comment]
**Update about the extension** - https://github.com/bumbeishvili/create-branch-from-issue

Awesome people added new functionalities to the script, now it's possible to choose source branch
![Branch](https://user-images.githubusercontent.com/6873202/104959403-9233ff80-59eb-11eb-8d8d-7329a1ce93b6.gif)

**Comment by ajRiverav** [comment]
potential alternative: https://github.com/marketplace/actions/create-issue-branch

**Comment by wasartin** [comment]
I find it hard to believe that the people at Github use their own project board. Because if they did, then they would have implemented this feature already.

**Comment by Ajisaputrars** [comment]
It is 2021 and still no such feature in GitHub?

**Comment by Shinigami92** [comment]
We should all move on to https://github.com/github/feedback/discussions instead of spamming here and not being heard

This issue should be closed and locked

**Comment by ajRiverav** [comment]
> We should all move on to [github/feedback/discussions](https://github.com/github/feedback/discussions) instead of spamming here and not being heard
> 
> This issue should be closed and locked

agreed. Do you have a link to an existing issue in github feedback? So that people here and up-vote it.

**Comment by Shinigami92** [comment]
No, someone should just create one and we can all vote on it

**Comment by ajRiverav** [comment]
there we go. 
I put it here https://github.com/github/feedback/discussions/3441

Let's all vote it up!

**Comment by krnlde** [comment]
Thank you guys!
